Add parent class blog post (#290)
epiverse-trace/epiverse-trace.github.io: posts/parent-class/index.qmd

---

> title: "Choosing the Right Parent for R Object Classes"

Reviewer: I think this title only partly captures the scope of the post.

I would suggest something like "Recommendations and good practices for creating custom S3 objects and methods in R".

Author: I think the suggestion is now not precise enough as we are likely to have other posts on specific good practices for S3. This is a specific areas of good practices for S3.

author:
  - name: "Hugo Gruson"
    orcid: "0000-0002-4094-1476"
date: "2024-06-24"
categories: [R, R package, object-oriented programming, S3, interoperability]
---

> [I have recently published](https://hugogruson.fr/posts/compa-tibble/) a [series of blog posts](https://epiverse-trace.github.io/posts/s3-generic/) on the reasons why one may want to start using object-oriented programming (and more specifically R S3 classes) to improve interoperability with other tools from the ecosystem.

Reviewer: Could you rather provide summary texts of the posts in square brackets?

For example, I have recently published about [making your inputs compatible with tibbles if they expect data.frames](URL) and [how to convert your functions to S3](URL) in a few steps.

Author: The goal here is to focus on the common topic across past posts, which helps justify this post.

But I think this is a good recommendation in general and will try to apply it wherever possible in the future.

But there are still questions I have not addressed directly, even if they may have been implicitly included sometimes: what makes a good object class? What good practices in class & function design can improve interoperability?

> As you can expect from these questions, this post may be slightly more subjective than the previous ones on this topic.

Reviewer: I think you can rather straightforwardly state that the post is to provide a rather subjective opinion on the matter.

I will argue that it is often a good strategy to inherit from existing standards classes, and to leverage this inheritance relationship as much as possible.

## Inherit from standard classes

> A peculiarity of R is the availability and centrality of `data.frame`s in the base language, whereas you need extra libraries for a similar functionality in most other languages (e.g., pandas in Python).

Author: ```suggestion
A peculiarity of R is the availability and centrality of `data.frame`s in the base language, whereas you need extra libraries for a similar functionality in most other languages (e.g., [pandas](https://pandas.pydata.org/) in Python).
```

> `data.frame` is one of the first "complex" (in the sense of [non-atomic](https://adv-r.hadley.nz/vectors-chap.html#atomic-vectors)) object most R learners will be exposed to and will develop a familiarity with. A good way to leverage this familiarity is to make your subclass a thin wrapper around `data.frame`s.

Author: ```suggestion
`data.frame` is one of the first "complex" (in the sense of [non-atomic](https://adv-r.hadley.nz/vectors-chap.html#atomic-vectors)) object most R learners will be exposed to and will develop a familiarity with. A good way to leverage this familiarity is to [make your subclass a thin wrapper around `data.frame`s](https://epiverse-trace.github.io/posts/extend-dataframes/).
```

> This means that not only users will be able to get started with your package faster because of this familiarity, but you will also immediately benefit for the huge ecosystem of functions and packages working on data.frames, such as the [tidyverse](https://tidyverse.org). If you want some examples, this is what we did in the [linelist](https://github.com/epiverse-trace/linelist), [pavo](https://github.com/rmaia/pavo), [scoringutils](https://github.com/epiforecasts/scoringutils), and [vaccineff](https://github.com/epiverse-trace/vaccineff) R packages.

Author: ```suggestion
This means that not only users will be able to get started with your package faster because of this familiarity, but you will also immediately benefit for the huge ecosystem of functions and packages working on `data.frame`s, such as the [tidyverse](https://tidyverse.org). If you want some examples, this is what we did in the [linelist](https://github.com/epiverse-trace/linelist), [pavo](https://github.com/rmaia/pavo), [scoringutils](https://github.com/epiforecasts/scoringutils), and [vaccineff](https://github.com/epiverse-trace/vaccineff) R packages.
```

> In some cases, the output is too complex to fit into a `data.frame`. Even in this situation, I would recommend inheriting from existing, well-established, classes for the same two reasons: familiarity and ecosystem. For example, for the [serofoi](https://github.com/epiverse-trace/serofoi) R package, [we have made the decision to inherit from `stanfit` objects, rather than a custom structure](https://github.com/epiverse-trace/serofoi/pull/117).

Author: There is already a link to the PR here so it may become messy. It's tangentially related so let's skip the link here.

## Rely on parent methods as much as possible

A follow up recommendation from inheriting from standard classes is to leverage their methods wherever possible.

> [One of the first](https://github.com/epiverse-trace/linelist/pull/60) [changes I made](https://github.com/epiverse-trace/linelist/pull/61) when becoming maintainer of the [linelist](https://github.com/epiverse-trace/linelist) package was to remove the `rename.linelist()` and `select.linelist()` methods. Indeed, they were, or could easily be, behaving identically as the parent `rename.data.frame()` and `select.data.frame()` methods. Rather than burdening the codebase and maintenance with an extra unnecessary method, it is much simpler and more robust to rely on the well-tested parent method. In fact, the [dplyr documentation](https://dplyr.tidyverse.org/reference/dplyr_extending.html) explicitly recommends only writing methods for a couple of standard functions (including `[.subclass()` and `names<-.subclass()`), which will enable the use of parent methods directly, rather than writing custom methods for each dplyr functions.

Author: As far as I can tell, both are correct and this is a matter of preference. I'm therefore resolving this document but please let me know if I missed something

> Similarly, many developers have the reflex to write a custom `print.subclass()` method as part of the method implementation. While it may be justified in some cases, it is also sometimes unnecessary extra fluff. My recommendation would be to evaluate carefully what benefit the custom method brings other defaulting to the parent method.

Author: ```suggestion
Similarly, many developers have the reflex to write a custom `print.subclass()` method as part of the method implementation. While it may be justified in some cases, it is also sometimes unnecessary. My recommendation would be to evaluate carefully what benefit the custom method brings other defaulting to the parent method.
```
